compiler/rustc_attr_parsing/src/lib.rs

```diff
@@ -45,14 +45,56 @@
 #![warn(unreachable_pub)]
 // tidy-alphabetical-end
 
+#[macro_use]
 mod attributes;
 mod context;
 pub mod parser;
 mod session_diagnostics;
 
-pub use attributes::*;
+pub use attributes::cfg::*;
+pub use attributes::util::{find_crate_name, is_builtin_attr, parse_version};
 pub use context::{AttributeParser, OmitDoc};
 pub use rustc_attr_data_structures::*;
-pub use util::{find_crate_name, is_builtin_attr, parse_version};
 
 rustc_fluent_macro::fluent_messages! { "../messages.ftl" }
+
+/// Finds attributes in sequences of attributes by pattern matching.
+///
+/// A little like `matches` but for attributes.
+///
+/// ```rust,ignore (illustrative)
+/// // finds the repr attribute
+/// if let Some(r) = find_attr!(attrs, AttributeKind::Repr(r) => r) {
+///
+/// }
+///
+/// // checks if one has matched
+/// if find_attr!(attrs, AttributeKind::Repr(_)) {
+///
+/// }
+/// ```
+#[macro_export]
+macro_rules! find_attr {
```

Comment:
I don't like this macro very much. Do you think, since you already have a macro with a list of all attribute parsers, that we could also generate a method per attribute that returns a list of attributes or an option of the attribute depending on whether it can have multiple results or not?

Similar to how we have different ways of accessing lang items

Reply:
hm, this is possible, though we need some identifier for it. That could be added to the other macro I guess. We need some (lowercase) identifier because the parser could parse one or several attributes so we don't always know what to choose from those, nor is it easy to access those symbols. I don't mind the macro myself, but if you think that's preferable I'm willing to try that

Reply:
oh I figured it out, it also checks whether item was already set before so we need to pass it in anyway. Still might prefer it that way.

Reply:
the problem is, what do you return out of this method? that's a more general problem, we don't have variant types. `get_repr()` should return something like `Option<Vec<ReprAttr>>` but other fns have different signatures. This is a more general problem with this entire approach, one that petrochenkov was a little afraid of I think. It's semantically nicer to have an enum, but it's hard to name a variant without its data, or data without its variant. That's why this is a macro, so you can use a pattern.

Reply:
My main worry is that it's not discoverable, so I'd end up writing 

```rust
for attr in tcx.get_all_attrs(did).filter_map(Attribute::parsed) {
    if let AttributeKind::Repr(r) = attr {
        ...
    }
}
```

because that's what I found when I looked on `TyCtxt`.

I see that generating the methods is impractical unless we have some common naming scheme where all the enum variants are `Repr(Repr)`

Reply:
mhm, you'd almost want like an "associated macro" for `tcx.get_attr!(pattern)`

Reply:
for your example specifically, a first solution would be to document how you're supposed to use it on `get_all_attrs`. Though by no means perfect...

Reply:
Maybe just name it `_raw`, then everyone will look at the docs to figure out what they should be doing

Reply:
mhm. Is there any proposal for `object.macro!()` which just invokes any macro in an impl block of a type and where self expands to `object`? It sounds sort of neat thinking about it...

Reply:
@WaffleLapkin says yes apparently, now I want it...

Reply:
I say meow

Reply:
so naming it _raw is also a bit iffy cause you do need to call it to then run `get_attr!()` on, like `get_attr!(tcx.get_attrs_raw(), pat => expr)`

Reply:
Maybe we can have a macro just for the pattern part, but a method on the attr list that is easily discoverable?
```rust
attrs.find_attr(|a| nonexhaustive_match!(a, AttributeKind::Repr(r) => r))
```
where `nonexhaustive_match!(e, pat => expr)` is
```rust
match e {
  pat => Some(expr),
  _ => None,
}
```
 The name `nonexhaustive_match` is terrible obviously, but up to that I think this is easier to discover and avoids a lot of the magic of `find_attr!`. (If we really want a shortcut for avoiding `is_some()`, that could be a separate method.)